include/oneapi/dpl/pstl/hetero/dpcpp/parallel_backend_sycl_reduce_then_scan.h

```diff
@@ -626,29 +714,87 @@ struct __gen_set_balanced_path
+            __rng1, __rng1_lower, __rng1_upper, __rng2, __rng2_lower, __rng2_upper, __i_elem, __comp);
 
         //Find balanced path for diagonal start
-        auto [__rng1_balanced_pos, __rng2_balanced_pos, __star_offset] =
-            __find_balanced_path_start_point(__rng1, __rng2, __rng1_pos, __rng2_pos, __comp);
+        auto [__rng1_balanced_pos, __rng2_balanced_pos, __star] = __find_balanced_path_start_point(
+            __rng1, __rng2, __rng1_pos, __rng2_pos, __rng1_lower, __rng2_lower, __rng2_upper);
 
         // Use sign bit to represent star offset. Temp storage is a signed type equal to the difference_type of the
         // input iterator range. The index will fit into the positive portion of the type, so the sign may be used to
         // indicate the star offset.
-        __rng1_temp_diag[__id] = __rng1_balanced_pos * (__star_offset ? -1 : 1);
+        __rng1_temp_diag[__id] =
+            oneapi::dpl::__par_backend_hetero::__encode_balanced_path_temp_data(__rng1_balanced_pos, __star);
+
+        return std::make_tuple(__rng1_balanced_pos, __rng2_balanced_pos, __star);
+    }
+
+    //Entry point for partitioning phase
+    template <typename _InRng, typename _IndexT>
+    void
+    __calc_partition_bounds(const _InRng& __in_rng, _IndexT __id) const
+    {
+        calc_and_store_balanced_path(__in_rng, __id, oneapi::dpl::__par_backend_hetero::__get_bounds_simple{});
+    }
+
+    // Entry point for reduce then scan reduce input
+    template <typename _InRng, typename _IndexT>
+    std::uint16_t
+    operator()(const _InRng& __in_rng, _IndexT __id, TempData& __temp_data) const
+    {
+        // First we must extract individual sequences from zip iterator because they may not have the same length,
+        // dereferencing is dangerous
+        auto __rng1 = std::get<0>(__in_rng.tuple());         // first sequence
+        auto __rng2 = std::get<1>(__in_rng.tuple());         // second sequence
+        auto __rng_tmp_diag = std::get<2>(__in_rng.tuple()); // temp diag sequence
+
+        _IndexT __rng1_balanced_pos = 0;
+        _IndexT __rng2_balanced_pos = 0;
+        bool __star = false;
 
-        _SizeType __eles_to_process = std::min(__diagonal_spacing - (__star_offset ? _SizeType{1} : _SizeType{0}),
-                                               __rng1.size() + __rng2.size() - (__i_elem - 1));
+        const auto __total_size = __rng1.size() + __rng2.size();
+        const bool __is_partitioned = __total_size >= __get_bounds.__partition_threshold;
+
+        if (__id * __diagonal_spacing >= __total_size)
+            return 0;
+        if (!__is_partitioned)
+        {
+            auto [__idx_rng1, __idx_rng2, __local_star] =
+                calc_and_store_balanced_path(__in_rng, __id, __get_bounds_simple{});
+            __rng1_balanced_pos = __idx_rng1;
+            __rng2_balanced_pos = __idx_rng2;
+            __star = __local_star;
+        }
+        else if (__id % __get_bounds.__tile_size != 0)
```

Comment:
I think a comment on how the balanced path calculations for each tile in the partition phase map to the balanced path computations in this kernel would be helpful. To me the details of the indexing can be a bit tricky to understand from just looking at the code.